trigger/src/main/scala/core/scheduler/JobScheduler.scala

```diff
@@ -0,0 +1,67 @@
+package core.scheduler
+
+import java.util.concurrent
+import java.util.concurrent.atomic.AtomicBoolean
+
+import core.persistance._
+import core.scheduler.executors.Executors
+import core.scheduler.sensors.Sensors
+import core.scheduler.utilities.{SchedulerConfig, SensorsConfig}
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+import scala.util.{Failure, Success}
+
+class JobScheduler(sensors: Sensors, executors: Executors, jobInstanceRepository: JobInstanceRepository) {
+  private val HEART_BEAT: Int = SchedulerConfig.getHeartBeat
+  val NUM_OF_PAR_TASKS: Int = SchedulerConfig.getMaxParallelJobs
+
+  private implicit val executionContext: ExecutionContextExecutor =
+    ExecutionContext.fromExecutor(concurrent.Executors.newFixedThreadPool(SensorsConfig.getThreadPoolSize))
+
+  private val isIsManagerRunning = new AtomicBoolean(false)
+  private var runningScheduler: Future[Unit] = Future.successful((): Unit)
+  private var runningSensors = Future.successful((): Unit)
+  private var runningEnqueue = Future.successful((): Unit)
+  private val runningJobs = mutable.Map.empty[Long, Future[Unit]]
+
+  def startManager(): Unit = {
+    isIsManagerRunning.set(true)
+    runningScheduler =
+      Future {
+        while(isIsManagerRunning.get()) {
+          removeFinishedJobs()
+          if(runningSensors.isCompleted)
+            runningSensors = sensors.processEvents()
+          if(runningJobs.size < NUM_OF_PAR_TASKS && runningEnqueue.isCompleted){
+            runningEnqueue = enqueueJobs(Math.max(0, NUM_OF_PAR_TASKS - runningJobs.size))
+          }
+          Thread.sleep(HEART_BEAT)
+        }
+      }
+    runningScheduler.onComplete {
+      case Success(_) =>
+        println("JobScheduler Stopped")
```

Comment:
maybe a logger?

Reply:
Yes of course. It will be implemented along with proper logging. It will take some time to Implement contextual logging. With term contextual logging I mean type of logging where I will place ID of jobInstance to all logs. Then I will be able pair logs with specific job run.